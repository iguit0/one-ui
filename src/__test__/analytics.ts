export function setCurrentPage(page: string) {
  const message = `Analytics: Setting current page ${page}`;
  (console.warn)(message);
}

export function logEvent(eventType: string, eventParams: any) {
  const message = `Analytics: Log event ${eventType}
${JSON.stringify(eventParams, null, 4)}`;
  (console.warn)(message);
}

export function setUserId(userId: string) {
  const message = `Analytics: Set user id ${userId}`;
  (console.warn)(message);
}
