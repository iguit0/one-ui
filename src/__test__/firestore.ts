import { doc, setDoc, updateDoc as _updateDoc } from "firebase/firestore";

export async function createDoc(firestore: any, path: string, data: any) {
  return await setDoc(doc(firestore, path), data);
}

export async function updateDoc<T extends Object = any>(
  firestore: any,
  path: string,
  data: Partial<T>
) {
  return await _updateDoc(doc(firestore, path), data as any);
}
