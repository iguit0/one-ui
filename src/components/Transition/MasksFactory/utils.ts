export default function buildGrid(
  amountOfVerticalSquares: number,
  howManyHorizontalSquare: number,
  ballSize: number
) {
  return new Array(howManyHorizontalSquare * amountOfVerticalSquares)
    .fill(undefined)
    .map((_, i) => {
      const modX = Math.floor(i / howManyHorizontalSquare);
      const modY = i - modX * howManyHorizontalSquare;
      const ballsize = ballSize * 1.1;
      const padding = (ballsize - ballSize) / 2;
      const x = modY * ballSize - padding;
      const y = modX * ballSize - padding;

      return {
        x,
        y,
        ballSize,
        column: modY,
        row: modX,
      };
    });
}

export function getTime(el: HTMLDivElement) {
  const _time =
    getComputedStyle(el.firstElementChild!).getPropertyValue(
      "--animation-speed-transition"
    ) || "250ms";
  const __time = _time.endsWith("ms")
    ? Number(_time.replace("ms", ""))
    : Number(_time.replace("s", "")) * 1000;
  return __time + 50;
}
