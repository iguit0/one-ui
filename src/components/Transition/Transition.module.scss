@import "../../assets/styles/index.scss";

.section {
  * {
    box-sizing: border-box;
  }
  position: relative;
  display: flex;
  align-items: center;
  > div {
    display: flex;
    flex-direction: column;
    min-height: inherit;
    min-width: 100%;
    animation-duration: var(--animation-speed-transition, $fast);
    animation-timing-function: var(--animation-timing-transition, initial);
    animation-fill-mode: forwards;
    transform-origin: inherit;
  }
}

div.entranceLeft {
  animation-name: entranceLeft;
}
div.exitLeft {
  animation-name: exitLeft;
}

@keyframes entranceLeft {
  0% {
    margin-left: -100%;
  }
  100% {
    margin-left: 0%;
  }
}
@keyframes exitLeft {
  0% {
    margin-left: 0%;
  }
  100% {
    margin-left: -100%;
  }
}

div.fadeOut {
  animation-name: fadeOut;
  position: absolute;
}

@keyframes fadeOut {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}

div.scaleIn {
  animation-fill-mode: forwards;
  animation-name: scaleIn;
}

@keyframes scaleIn {
  0% {
    transform: scale(0);
  }
  100% {
    transform: scale(1);
  }
}

div.scaleOut {
  animation-fill-mode: initial;
  animation-name: scaleOut;
}

@keyframes scaleOut {
  0% {
    position: absolute;
    transform: scale(1);
  }
  100% {
    position: absolute;
    transform: scale(0);
  }
}

div.fadeIn {
  animation-name: fadeIn;
}

@keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

@keyframes fadeInDelayed {
  0% {
    opacity: 0;
  }
  50% {
    opacity: 0.1;
  }
  100% {
    opacity: 1;
  }
}
@keyframes fadeOutDelayed {
  0% {
    opacity: 1;
  }
  50% {
    opacity: 0.1;
  }
  100% {
    opacity: 0;
  }
}

.fadeInDelayed {
  animation-name: fadeInDelayed;
}

.fadeOutAbsolute {
  position: absolute;
  animation-name: fadeOutDelayed;
}

@keyframes flip {
  0% {
    transform: rotateX(0deg);
  }

  25% {
    transform: rotateX(36deg);
  }

  50% {
    transform: rotateX(90deg);
  }

  75% {
    transform: rotateX(90deg);
  }

  100% {
    transform: rotateX(90deg);
  }
}
@keyframes flipIn {
  0% {
    transform: rotateX(0deg);
    opacity: 0;
  }

  25% {
    transform: rotateX(0deg);
  }

  50% {
    transform: rotateX(90deg);
    opacity: 0;
  }
  51% {
    opacity: 1;
  }

  75% {
    transform: rotateX(36deg);
  }

  100% {
    transform: rotateX(0deg);
  }
}
.flipBottomUpOut {
  transform-origin: 50% 50% !important;
  position: absolute !important;
  animation-name: flip;
  animation-timing-function: linear !important;
}
.flipBottomUpIn {
  transform-origin: 50% 50% !important;
  animation-name: flipIn;
  animation-timing-function: linear !important;
}
