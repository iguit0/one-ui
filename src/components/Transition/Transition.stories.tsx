import React from "react";
import Text from "../Text";
import _Transition from "./Transition";

export default {
  component: _Transition,
  title: "Transitor de elementos",
};

export const InitialImplementation = (args: any) => (
  <>
    <_Transition {...args}>
      <div style={{ backgroundColor: "red" }}>
        <Text style={{ padding: 48, textAlign: "center" }} type="title">
          Etapa 1
        </Text>
      </div>
      <div style={{ backgroundColor: "green" }}>
        <Text style={{ padding: 48, textAlign: "center" }} type="title">
          Etapa 2
        </Text>
      </div>
      <div style={{ backgroundColor: "blue" }}>
        <Text style={{ padding: 48, textAlign: "center" }} type="title">
          Etapa 3
        </Text>
      </div>
      <div style={{ backgroundColor: "yellow" }}>
        <Text style={{ padding: 48, textAlign: "center" }} type="title">
          Etapa 4
        </Text>
      </div>
    </_Transition>
    <h1>Controle o elemento atual pelos controles abaixo, existem 4 etapas</h1>
  </>
);
InitialImplementation.args = {
  step: 0,
} as Partial<React.ComponentProps<typeof _Transition>>;
