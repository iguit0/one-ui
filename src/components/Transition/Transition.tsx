import React, {
  createRef,
  PropsWithChildren,
  useEffect,
  useRef,
  useState,
} from "react";
import Styles from "./Transition.module.scss";

export type TransitionProps = {
  step: number;
  className?: string;
  contentStyle?: React.CSSProperties;
  contentClassName?: string;
  children: (React.ReactElement | undefined)[];
  onDiscardStep?: (discardedKey: React.Key) => void;
  lockTransitionWidth?: boolean;
  transitionType?: TransitionAnimationTypes;
};

function TransitionClasses(
  type: NonNullable<TransitionProps["transitionType"]>
): {
  /**
   * Applied to the element that was not visible and is now entering the screen
   */
  backward: {
    elementEntering: string;
    elementExiting: string;
  };
  /**
   * Applied to the element that was visible and now is exiting the screen
   */
  forward: {
    elementEntering: string;
    elementExiting: string;
  };
} {
  switch (type) {
    case TransitionAnimationTypes.SLIDE:
      return {
        backward: {
          elementEntering: Styles.entranceLeft,
          elementExiting: "",
        },
        forward: {
          elementEntering: "",
          elementExiting: Styles.exitLeft,
        },
      };
    case TransitionAnimationTypes.POP_FROM_CLICK_ORIGIN:
      return {
        backward: {
          elementEntering: Styles.fadeIn,
          elementExiting: Styles.scaleOut,
        },
        forward: {
          elementEntering: Styles.scaleIn,
          elementExiting: Styles.fadeOut,
        },
      };
  }
}

/**
 * Handles the transition between multiple children and recycling of elements
 **/
export default function Transition({
  step,
  contentStyle,
  contentClassName = "",
  children,
  className,
  onDiscardStep,
  lockTransitionWidth = false,
  transitionType = TransitionAnimationTypes.SLIDE,
}: TransitionProps) {
  const preTransitionDetails = useRef<{
    transformOrigin: `${number}% ${number}%` | `initial`;
  }>({
    transformOrigin: "initial",
  });
  const containerRef = useRef<HTMLElement>(null);
  const [screensStack, setScreensStack] = useState([
    <div
      data-testid="transition-container"
      key={String(children[step]?.key || step)}
      style={contentStyle}
      className={contentClassName}
    >
      {children[step]}
    </div>,
  ]);

  const prevStep = useRef(step);
  const prevKey = useRef(children[step]?.key);

  useEffect(() => {
    const transitionClasses = TransitionClasses(transitionType);

    if (
      prevKey.current !== null &&
      children[step]?.key === prevKey.current // I'm rendering the same screen
    ) {
      return () => {
        prevStep.current = step;
      };
    }

    const key = children[step]?.key || step;
    console.warn(key);

    if (prevStep.current !== step && lockTransitionWidth)
      containerRef.current!.style.width = `${
        containerRef.current!.clientWidth
      }px`;

    if (prevStep.current > step) {
      const stepToRemove = prevStep.current;
      const prevKeyToRemove = prevKey.current || stepToRemove;
      setScreensStack((screensBeforeChangingStep) => {
        const enteringScreenRef = createRef<HTMLDivElement>();
        const [firstNextScreen, ...restOfScreens] =
          screensBeforeChangingStep.filter((a) => a.key !== String(key));
        const clonedFirst = React.cloneElement(firstNextScreen, {
          "data-testid": "transition-container",
          style: {
            ...contentStyle,
            ...preTransitionDetails.current,
          },
          className: `${transitionClasses.backward.elementExiting} ${
            firstNextScreen.props.className?.replace(
              transitionClasses.backward.elementEntering,
              ""
            ) || ""
          }`,
        });
        return [
          <div
            ref={enteringScreenRef}
            data-testid="transition-container"
            key={key}
            className={`${transitionClasses.backward.elementEntering} ${contentClassName}`}
            style={{
              ...contentStyle,
              ...preTransitionDetails.current,
            }}
            onAnimationEnd={() => {
              setScreensStack((screensAfterTheCurrentStepEntered) => {
                if (onDiscardStep) onDiscardStep(prevKeyToRemove);
                return screensAfterTheCurrentStepEntered.filter(
                  (s) => s.key !== String(prevKeyToRemove)
                );
              });
            }}
          >
            {children[step]}
          </div>,
          clonedFirst,
          ...restOfScreens,
        ];
      });
    } else if (prevStep.current < step) {
      const stepToDelete = prevStep.current;
      const prevKeyToRemove = prevKey.current || stepToDelete;
      setScreensStack((screensBeforeChangingStep) => {
        const lastIndex = screensBeforeChangingStep.length - 1;
        const lastScreen = screensBeforeChangingStep[lastIndex];
        const nextScreenRef = createRef<HTMLDivElement>();
        const clonedLast = React.cloneElement(lastScreen, {
          "data-testid": "transition-container",
          style: {
            ...contentStyle,
            ...preTransitionDetails.current,
          },
          className: `${transitionClasses.forward.elementExiting} ${
            lastScreen.props.className?.replace(
              transitionClasses.backward.elementEntering,
              ""
            ) || ""
          }`,
          onAnimationEnd: () => {
            nextScreenRef.current!.classList.remove(
              transitionClasses.forward.elementEntering
            );
            if (onDiscardStep) onDiscardStep(prevKeyToRemove);
            setScreensStack((screensAfterTheCurrentStepEntered) => {
              return screensAfterTheCurrentStepEntered.filter((s) => {
                return s.key !== String(prevKeyToRemove);
              });
            });
          },
        });

        return [
          ...screensBeforeChangingStep.slice(0, lastIndex),
          clonedLast,
          <div
            ref={nextScreenRef}
            data-testid="transition-container"
            key={key}
            style={{
              ...contentStyle,
              ...preTransitionDetails.current,
            }}
            className={`${contentClassName} ${transitionClasses.forward.elementEntering}`}
          >
            {children[step]}
          </div>,
        ];
      });
    }
    return () => {
      prevStep.current = step;
      prevKey.current = children[step]?.key;
    };
  }, [step]);

  useEffect(() => {
    if (screensStack.length === 1 && lockTransitionWidth)
      containerRef.current!.style.width = "";
  }, [screensStack.length !== 1]);

  return (
    <>
      <section
        onClick={({
          currentTarget: { offsetTop, offsetLeft, clientWidth, clientHeight },
          clientX,
          clientY,
        }) => {
          const offsetX = clientX - offsetLeft;
          const offsetY = clientY - offsetTop;
          const percentX = (offsetX * 100) / clientWidth;
          const percentY = (offsetY * 100) / clientHeight;
          preTransitionDetails.current.transformOrigin = `${percentX}% ${percentY}%`;
        }}
        data-testid="transition-controller"
        ref={containerRef}
        className={`${Styles.section} ${className}`}
      >
        {screensStack}
      </section>
    </>
  );
}

export enum TransitionAnimationTypes {
  SLIDE,
  POP_FROM_CLICK_ORIGIN,
}
