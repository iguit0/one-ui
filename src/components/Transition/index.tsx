export { default, TransitionAnimationTypes } from "./Transition";
