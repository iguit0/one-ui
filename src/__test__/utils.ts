// @ts-nocheck

import { Firestore } from "firebase/firestore";
import { FirebaseStorage } from "firebase/storage";
import {
  RulesTestEnvironment,
  initializeTestEnvironment,
} from "@firebase/rules-unit-testing";

// import { ChildProcess, execSync, spawn, SpawnSyncReturns } from "child_process";
export function WaitTimeout(ml = 200) {
  return new Promise<void>((r) => {
    setTimeout(() => {
      r();
    }, ml);
    if (window.Cypress) return;
    else if (jest.isMockFunction(setTimeout)) jest.advanceTimersByTime(ml);
  });
}

class FirebaseEmulatorInterface {
  private process!: any;

  private killRelatedPorts() {
    for (let port of [4000, 4400, 8055, 8080]) {
      try {
        if (window.Cypress)
          cy.exec(`yarn kill-port ${port}`, { failOnNonZeroExit: true })
        else
          require("child_process").execSync(`kill -9 $(lsof -t -i:${port})`, {
            stdio: "ignore",
          });
      } catch (e) {
        console.error("Error when killing port", port, e);
      }
    }
  }

  start(fakeProjectName: string, databaseToImport?: string = "") {
    if (sessionStorage.getItem("last-database") === databaseToImport)
      return;
    this.killRelatedPorts();
    if (window.Cypress) {
      const command = `yarn start-emulator ${fakeProjectName} ${databaseToImport}`;
      cy.exec(command, {
        timeout: 60000, // Instant fail
      });
      sessionStorage.setItem("last-database", databaseToImport)
    } else {
      this.process = require("child_process").spawn(
        "firebase",
        ["emulators:start", "-P", "whz-test"],
        {
          stdio: "pipe",
        }
      );

      return new Promise<void>((r, rej) => {
        let fullOutput = "";
        const timeoutSeconds = 20;
        const unexpectedTimeout = setTimeout(() => {
          rej(
            new Error(`Waited for 20 seconds and didn't received ok from emulator.
        
Below is the full output:

--------EMULATOR START OUTPUT--------

${fullOutput}

--------EMULATOR START OUTPUT--------`)
          );
        }, timeoutSeconds * 1000);
        this.process.stdout!.on("data", (txt) => {
          const output = txt.toString();
          fullOutput += output;

          if (output.includes("All emulators ready!")) {
            clearTimeout(unexpectedTimeout);
            setTimeout(() => {
              r();
            }, 2000);
          }
        });
      });
    }
  }
  async stop(immediate?: boolean, timeoutSec?: number) {
    const write = window.Cypress ? console.log : process.stdout.write;
    const timeout = (timeoutSec || 10) * 1000;
    const kill = (r: Function) => {
      if (window.Cypress) this.killRelatedPorts();
      else this.process.kill("SIGINT");

      setTimeout(() => {
        r();
      }, 2000);
    };
    if (immediate) {
      return new Promise((res) => {
        kill(res);
      });
    } else {
      const clear = (message: string) => {
        if (window.Cypress) return;
        write(new Array(message.length).fill("\r").join(""));
        write(new Array(message.length).fill(" ").join(""));
      };
      return new Promise((res) => {
        let remaining = timeout / 1000;
        let message!: string;
        const interval = setInterval(() => {
          if (message) clear(message);
          message = `Killing emulator in ${remaining} seconds`;
          write(message);

          remaining -= 1;
          if (remaining === 0) {
            kill(res);
            clear(message);
            write("Emulator killed");
            clearInterval(interval);
          }
        }, 1000);
      });
    }
  }
}

export const EmulatorController = new FirebaseEmulatorInterface();

export function cyPromise(someAssertion: { then: Function }) {
  return new Promise((r) => {
    someAssertion.then(r);
  });
}

export async function initTestFirebase(
  fakeProjectId: string,
  setupEnv?: (
    ctx: Firestore,
    storageCtx: FirebaseStorage,
    testEnv: RulesTestEnvironment
  ) => Promise<void>
): Promise<RulesTestEnvironment> {
  const testEnv = await initializeTestEnvironment({
    projectId: fakeProjectId,
    firestore: {
      host: "localhost",
      port: 8080,
    },
    storage: {
      host: "localhost",
      port: 9199,
    },
  });

  await testEnv.withSecurityRulesDisabled(async (ctx) => {
    const firestore = ctx.firestore({
      experimentalForceLongPolling: true,
    });

    const storage = ctx.storage() as any;
    if (setupEnv) await setupEnv(firestore, storage, testEnv);
  });
  return testEnv;
}
