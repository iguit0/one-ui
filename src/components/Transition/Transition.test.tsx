import React from "react";
import { render } from "@testing-library/react";

import { InitialImplementation as Component } from "./Transition.stories";
import { WaitTimeout } from "../../__test__/utils";

function renderScreen(props: React.ComponentProps<typeof Component>) {
  return render(<Component {...props} />);
}

it("Should at least render :)", () => {
  renderScreen({ step: 0 });
});

it("Should not break this crazy logic", async () => {
  const wrapper = render(<Component step={0} />);

  wrapper.rerender(<Component step={1} />);
  wrapper.rerender(<Component step={2} />);
  wrapper.rerender(<Component step={3} />);
  wrapper.rerender(<Component step={2} />);
  const [el3WhenGoinBack, el1, el2, el4] = wrapper.getAllByTestId(
    "transition-container"
  ) as HTMLDivElement[];
  expect(wrapper.container.textContent).toMatchSnapshot();
  const event = new Event("animationend");
  event.initEvent("animationend", true, true);
  el1.dispatchEvent(event);
  expect(wrapper.container.textContent).toMatchSnapshot();
  el2.dispatchEvent(event);
  expect(wrapper.container.textContent).toMatchSnapshot();
  el3WhenGoinBack.dispatchEvent(event);
  expect(wrapper.container.textContent).toMatchSnapshot();
});

it("Should lock the current width when transitioning", async () => {
  const wrapper = render(<Component lockTransitionWidth step={0} />);
  const el = wrapper.getByTestId("transition-controller");
  expect(el.style.width).toEqual("");
  wrapper.rerender(<Component lockTransitionWidth step={1} />);
  expect(el.style.width).toEqual("0px");

  const [el1] = wrapper.getAllByTestId(
    "transition-container"
  ) as HTMLDivElement[];
  const event = new Event("animationend");
  event.initEvent("animationend", true, true);
  el1.dispatchEvent(event);
  await WaitTimeout();
  expect(el.style.width).toEqual("");
});
