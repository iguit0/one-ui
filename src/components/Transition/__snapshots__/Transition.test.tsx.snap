// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should not break this crazy logic 1`] = `"Etapa 3Etapa 1Etapa 2Etapa 4Controle o elemento atual pelos controles abaixo, existem 4 etapas"`;

exports[`Should not break this crazy logic 2`] = `"Etapa 3Etapa 2Etapa 4Controle o elemento atual pelos controles abaixo, existem 4 etapas"`;

exports[`Should not break this crazy logic 3`] = `"Etapa 3Etapa 4Controle o elemento atual pelos controles abaixo, existem 4 etapas"`;

exports[`Should not break this crazy logic 4`] = `"Etapa 3Controle o elemento atual pelos controles abaixo, existem 4 etapas"`;
